### PR TITLE
Upgrade to Sentry SDK 5

### PR DESCRIPTION
**Pull request by verglor**
Upgrade to the latest stable version of sentry-java to enable performance monitoring.

### CODE REVIEW

**Comment by musketyr** [comment]
Hi, @verglor. Thank you for this PR. This is quite a radical change. I know as I have already done the migration for the Micronaut version. I'm adding @jvdrean to help me with the review but the current state on our side is that we're leaving the Grails ecosystem in upcoming months and we'll be discontinuing our Grails plugins. But unless we kill our last Grails app we need to keep our plugins stable.

I would rather suggest you to release your own fork of the plugin on Maven Central.

I will need to dig deeper in the new concepts you have introduces and maybe carry some of them into https://github.com/agorapulse/micronaut-log4aws which is more or less successor of this plugin.

**Comment by verglor** [comment]
Hi @musketyr, 
maybe there can be `sentry-5` or `v2` branch with separate `2.x.x` releases so that it is available at `org.grails.plugins:sentry:2.x.x` but maintenance and releases of old version are still possible ?
I wouldn't like to create a completely new plugin if it's not necessary.
Do you plan to transfer this plugin to a new maintainer once you discontinue it ?

**Comment by musketyr** [comment]
Hi @verglor, transferring the plugin to the new owner is definitely an option if you or someone else would like to do so. 

The current situation is that there is even no pipeline to release set up since the Bintray's EOL. With the other plugins we switched to [Jitpack](https://github.com/agorapulse/grails-aws-sdk#installation) for releases as a temporary solution so we don't have to change the coordinates and setup all the signing and other requires but the best solution would be to release the plugin under new coordinates to Maven Central. So it would be kind of  _new plugin_ anyway.

File: README.md [hunk 1]
```diff
@@ -60,34 +60,30 @@ You can also set the server name, but it is recommended to don't set this config
...
-                MODULE1: [com.company.services.module1, com.company.controllers.module1]
-                MODULE2: [com.company.services.module2, com.company.controllers.module2]
-                MODULE3: [com.company.services.module3, com.company.controllers.module3]
-            logClassName: true
             logHttpRequest: true
             disableMDCInsertingServletFilter: true
+            inAppIncludes:
+              - my.app.package.one
+              - my.app.package.two
+            tracesSampleRate: 0.5
```
Comment (leni-kirilov) [inline README.md:77]: Is this a recommended default trace sample rate ? Should we add it the to the README ?

File: grails-app/controllers/grails/plugin/sentry/SentryTracingInterceptor.groovy [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package grails.plugin.sentry
+
+import org.springframework.web.servlet.HandlerMapping
+
+class SentryTracingInterceptor {
```
Comment (leni-kirilov) [inline grails-app/controllers/grails/plugin/sentry/SentryTracingInterceptor.groovy:5]: This interceptor is the same we have in our grails app. Would we be able to remove it in the grails-app ?

File: grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import io.sentry.Sentry
+import io.sentry.SentryTraceHeader
+
+class SentryTagLib {
+
+    static final namespace = 'sentry'
+
+    SentryConfig sentryPluginConfig
+
+    def traceMeta = { attrs, body ->
```
Comment (leni-kirilov) [inline grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy:19]: maybe some javadoc would be nice (in this class as in general)

File: grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        String linkPrefix = sentryPluginConfig.linkPrefix
+        if (sentryTraceId && linkPrefix) {
+            out << """<a href="${linkPrefix}/performance/trace/${sentryTraceId}" target="_blank class="sentry-trace-link">"""
+            out << (body() ?: """
+                <svg style="vertical-align: text-bottom; margin: 0 8px;" fill="currentColor" viewBox="0 0 16 16" height="16px" width="16px">
+                    <path d="M15.8,14.57a1.53,1.53,0,0,0,0-1.52L9.28,1.43a1.46,1.46,0,0,0-2.56,0L4.61,5.18l.54.32A10.43,10.43,0,0,1,8.92,9.39a10.84,10.84,0,0,1,1.37,4.67H8.81a9.29,9.29,0,0,0-1.16-3.91A9,9,0,0,0,4.41,6.81L3.88,6.5,1.91,10l.53.32a5.12,5.12,0,0,1,2.42,3.73H1.48a.25.25,0,0,1-.21-.12.24.24,0,0,1,0-.25L2.21,12a3.32,3.32,0,0,0-1.07-.63L.2,13.05a1.53,1.53,0,0,0,0,1.52,1.46,1.46,0,0,0,1.28.76H6.13V14.7a6.55,6.55,0,0,0-.82-3.16,6.31,6.31,0,0,0-1.73-2l.74-1.32a7.85,7.85,0,0,1,2.26,2.53,8,8,0,0,1,1,3.92v.63h3.94V14.7A12.14,12.14,0,0,0,10,8.75a11.8,11.8,0,0,0-3.7-4l1.5-2.67a.24.24,0,0,1,.42,0l6.52,11.63a.24.24,0,0,1,0,.25.24.24,0,0,1-.21.12H13c0,.43,0,.85,0,1.27h1.53a1.46,1.46,0,0,0,1.28-.76"></path>
+                </svg>
```
Comment (leni-kirilov) [inline grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy:42]: Maybe this svg can be extracted to its own file? I don't see anything dynamic in it
Edit: this is sentry's logo. We could also reference their logo by URL ?

File: README.md [hunk 5]
```diff
@@ -100,36 +96,48 @@ Check [Sentry-java](https://github.com/getsentry/sentry-java) documentation to c
...
+You can also enable [JDBC tracing](https://docs.sentry.io/platforms/java/performance/instrumentation/jdbc/) by setting config `traceJDBC` to `true`.
+
+To enable [distributed tracing](https://docs.sentry.io/product/sentry-basics/tracing/distributed-tracing/) you need to [connect services](https://docs.sentry.io/platforms/java/performance/connect-services/) (frontend) 
+by adding `<sentry:traceMeta />` to `<head>` of your gsp view. 
+
+If you configure `linkPrefix` you can also add `<sentry:traceLink />` to your gsp view which displays direct link to transaction in sentry application.
+
 # Latest releases
 
 * 2018-07-16 **V11.7.25** : upgrade Sentry java lib to 1.7.25 + stack trace sanitizer
```
Comment (leni-kirilov) [inline README.md:143]: Latest release entry is needed ;) Possibly if @musketyr doesn't want this new version in their repo, they can at least add a link to this fresher version in their README.

File: src/main/groovy/grails/plugin/sentry/SentryConfig.groovy [hunk 6]
```diff
@@ -95,6 +93,25 @@ class SentryConfig {
             sanitizeStackTrace = true
         }
 
+        if (config.inAppIncludes) {
+            if (config.inAppIncludes instanceof List) {
+                inAppIncludes = config.inAppIncludes as List
+            }
+            if (config.inAppIncludes instanceof String) {
+                inAppIncludes = (config.inAppIncludes as String).split(",") as List
```
Comment (leni-kirilov) [inline src/main/groovy/grails/plugin/sentry/SentryConfig.groovy:101]: if we support a csv, do we care about whitespace trimming?

File: src/main/resources/spy.properties [hunk 7]
```diff
@@ -0,0 +1,2 @@
+modulelist=com.p6spy.engine.spy.P6SpyFactory
+useNanoTime=true
```
Comment (leni-kirilov) [inline src/main/resources/spy.properties:2]: Why do we want nanotime? I think it's not guaranteed better accuracy on most cpus/oss

File: src/test/groovy/grails/plugin/sentry/SanityIntegrationSpec.groovy [hunk 8]
```diff
@@ -4,43 +4,38 @@ import ch.qos.logback.classic.Logger
...
-                post("/api/123/store/") {
+                post("/api/123/envelope/") {
                     called 1
                     responder {
                         code 200
                     }
                 }
             }
             server.start()
+            println "Ersatz server started at localhost:${server.httpPort}"
```
Comment (leni-kirilov) [inline src/test/groovy/grails/plugin/sentry/SanityIntegrationSpec.groovy:34]: prinltn ?

File: src/test/groovy/grails/plugin/sentry/SentryTagLibSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    void "traceLink (sentry enabled with linkPrefix and body)"() {
+        given:
+        tagLib.sentryPluginConfig.getLinkPrefix() >> 'http://prefix'
+
+        SentryOptions options = new SentryOptions()
+        options.dsn = "http://key@localhost/42"
+        options.tracesSampleRate = 1.0
+        Sentry.init(options)
+        Sentry.startTransaction('name', 'operation', true)
+        String sentryTraceId = Sentry.currentHub.span?.toSentryTrace()?.traceId
```
Comment (leni-kirilov) [inline src/test/groovy/grails/plugin/sentry/SentryTagLibSpec.groovy]: This setup repeats in most tests (with only the transaction part seems to be different) . I suggest extracting it to a util method

**Review by leni-kirilov — COMMENTED** [review]
I'll refrain from approving this PR for an unfamiliar plugin for me. 
I added a few comments nevertheless. 
Looks solid, IMHO ! Good work!

File: README.md (hunk 1 above)
Comment (verglor) [reply README.md:77]: No actually, just random example ;-)

File: grails-app/controllers/grails/plugin/sentry/SentryTracingInterceptor.groovy (hunk 2 above)
Comment (verglor) [reply grails-app/controllers/grails/plugin/sentry/SentryTracingInterceptor.groovy:5]: This is default, that can be overwritten. It's still useful for performance monitoring working out of the box. Much better approach would be [providing custom TransactionNameProvider](https://github.com/getsentry/sentry-java/pull/1861) bean, but that is possible only in sentry-java v6 which is still beta.

File: src/main/resources/spy.properties (hunk 7 above)
Comment (verglor) [reply src/main/resources/spy.properties:2]: Many queries are sub millisecond and are displayed as just 1ms then.

File: src/test/groovy/grails/plugin/sentry/SanityIntegrationSpec.groovy (hunk 8 above)
Comment (verglor) [reply src/test/groovy/grails/plugin/sentry/SanityIntegrationSpec.groovy:34]: Yeah, logging is not actually displayed during test, this is still visible only with 
```
test {
    testLogging {
        showStandardStreams = true
    }
}
```

File: src/main/groovy/grails/plugin/sentry/SentryConfig.groovy (hunk 6 above)
Comment (verglor) [reply src/main/groovy/grails/plugin/sentry/SentryConfig.groovy:101]: This is the way how lists are handled in this class so I used the same for consistency. Actually I don't like it much at all and would rather use standard `@ConfigurationProperties(prefix = "grails.plugin.sentry")`, but there was problem with accessing this bean in [SentryGrailsPlugin.doWithSpring()](https://github.com/agorapulse/grails-sentry/blob/781a7d7c28eb2aea2bcfac6efdfd3e2c85e1179b/src/main/groovy/grails/plugin/sentry/SentryGrailsPlugin.groovy#L63).

File: README.md (hunk 5 above)
Comment (musketyr) [reply README.md:143]: I will happily do so. I don't think we're going to adopt the version 5 for our Grails apps.

File: grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy (hunk 4 above)
Comment (verglor) [reply grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy:42]: This was actually took from sentry page while looking for url of their logo. I think it's fine to have it embedded sparing one request and dealing with plugin assets.

File: README.md (hunk 5 above)
Comment (verglor) [reply README.md:143]: I didn't know when and what version will be released and what all will it contain. I assumed the entry would be created just before release.

File: src/test/groovy/grails/plugin/sentry/SentryTagLibSpec.groovy (hunk 9 above)
Comment (verglor) [reply src/test/groovy/grails/plugin/sentry/SentryTagLibSpec.groovy]: 55d487f5

File: grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy (hunk 4 above)
Comment (leni-kirilov) [reply grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy:42]: OK about the URL and requests, but the code and tests would more readable, too

File: grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy (hunk 3 above)
Comment (verglor) [reply grails-app/taglib/grails/plugin/sentry/SentryTagLib.groovy:19]: 724873e

**Comment by verglor** [comment]
Closed in favor of https://github.com/pmedianetwork/grails-sentry/pull/1

**Pull request closed without merging**